quic/s2n-quic-core/src/inet/ipv4.rs

```diff
@@ -305,6 +314,330 @@ impl From<IpV4Address> for [u8; IPV4_LEN] {
+        &self.flag_fragment
+    }
+
+    #[inline]
+    pub fn flag_fragment_mut(&mut self) -> &mut FlagFragment {
+        &mut self.flag_fragment
+    }
+
+    #[inline]
+    pub const fn ttl(&self) -> &u8 {
+        &self.ttl
+    }
+
+    #[inline]
+    pub fn ttl_mut(&mut self) -> &mut u8 {
+        &mut self.ttl
+    }
+
+    #[inline]
+    pub const fn protocol(&self) -> &ip::Protocol {
+        &self.protocol
+    }
+
+    #[inline]
+    pub fn protocol_mut(&mut self) -> &mut ip::Protocol {
+        &mut self.protocol
+    }
+
+    #[inline]
+    pub const fn checksum(&self) -> &U16 {
+        &self.checksum
+    }
+
+    #[inline]
+    pub fn checksum_mut(&mut self) -> &mut U16 {
+        &mut self.checksum
+    }
+
+    #[inline]
+    pub const fn source(&self) -> &IpV4Address {
+        &self.source
+    }
+
+    #[inline]
+    pub fn source_mut(&mut self) -> &mut IpV4Address {
+        &mut self.source
+    }
+
+    #[inline]
+    pub const fn destination(&self) -> &IpV4Address {
+        &self.destination
+    }
+
+    #[inline]
+    pub fn destination_mut(&mut self) -> &mut IpV4Address {
+        &mut self.destination
+    }
+}
+
+define_inet_type!(
+    pub struct Vihl {
+        value: u8,
+    }
+);
+
+impl fmt::Debug for Vihl {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("Vihl")
+            .field("version", &self.version())
+            .field("header_len", &self.header_len())
+            .finish()
+    }
+}
+
+impl Vihl {
+    #[inline]
+    pub fn version(&self) -> u8 {
+        self.value >> 4
+    }
+
+    #[inline]
+    pub fn set_version(&mut self, value: u8) -> &mut Self {
+        self.value = value << 4 | (self.value & 0x0F);
+        self
+    }
+
+    #[inline]
+    pub fn header_len(&self) -> u8 {
+        self.value & 0x0F
+    }
+
+    #[inline]
+    pub fn set_header_len(&mut self, value: u8) -> &mut Self {
+        self.value = (self.value & 0xF0) | (value & 0x0F);
+        self
+    }
+}
+
+define_inet_type!(
+    pub struct Tos {
+        value: u8,
+    }
+);
+
+impl Tos {
+    /// Differentiated Services Code Point
+    #[inline]
+    pub fn dscp(&self) -> u8 {
+        self.value >> 2
+    }
+
+    #[inline]
+    pub fn set_dscp(&mut self, value: u8) -> &mut Self {
+        self.value = (value << 2) | (self.value & 0b11);
+        self
+    }
+
+    #[inline]
+    pub fn ecn(&self) -> ExplicitCongestionNotification {
+        ExplicitCongestionNotification::new(self.value & 0b11)
+    }
+
+    #[inline]
+    pub fn set_ecn(&mut self, ecn: ExplicitCongestionNotification) -> &mut Self {
+        self.value = (self.value & !0b11) | ecn as u8;
+        self
+    }
+}
+
+impl fmt::Debug for Tos {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("ipv4::Tos")
+            .field("dscp", &self.dscp())
+            .field("ecn", &self.ecn())
+            .finish()
+    }
+}
+
+define_inet_type!(
```

Comment:
I'm relatively unfamiliar with the ipv4 networking stack, but this type feels particularly dense. It looks like the most significant 3 bits are reserved for `reserved`, `DF` and `more fragments` fields, but the `fragment_offset` and `set_fragment_offset` aren't immediately clear to me. I can tell you what bits they return, but I don't really understand why they are used for.